pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -92,61 +92,55 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
 		for _, attachedVolume := range rc.actualStateOfWorld.GetAttachedVolumes() {
 			if !rc.desiredStateOfWorld.VolumeExists(
 				attachedVolume.VolumeName, attachedVolume.NodeName) {
-				// Volume exists in actual state of world but not desired
-
-				// Mark desire to detach
-				timeElapsed, err := rc.actualStateOfWorld.MarkDesireToDetach(attachedVolume.VolumeName, attachedVolume.NodeName)
+				// Set the detach request time
+				elapsedTime, err := rc.actualStateOfWorld.SetDetachRequestTime(attachedVolume.VolumeName, attachedVolume.NodeName)
 				if err != nil {
-					glog.Errorf("Unexpected error actualStateOfWorld.MarkDesireToDetach(): %v", err)
+					glog.V(4).Infof("Cannot trigger detach because %v", err)
+					continue
+				}
+				// Check whether timeout has reached the maximum waiting time
+				timeout := elapsedTime > rc.maxWaitForUnmountDuration
+				// Check whether volume is still mounted. Skip detach if it is still mounted unless timeout
+				if attachedVolume.MountedByNode && !timeout {
+					glog.V(4).Infof("Cannot trigger detach for volume %q on node %q because volume is still mounted",
+						attachedVolume.VolumeName,
+						attachedVolume.NodeName)
+					continue
 				}
 
+				// Before triggering volume detach, mark volume as detached and update the node status
+				// If it fails to update node status, skip detach volume
+				rc.actualStateOfWorld.MarkDesireToDetachVolume(attachedVolume.VolumeName, attachedVolume.NodeName)
+
 				// Update Node Status to indicate volume is no longer safe to mount.
 				err = rc.nodeStatusUpdater.UpdateNodeStatuses()
 				if err != nil {
 					// Skip detaching this volume if unable to update node status
-					glog.Infof("UpdateNodeStatuses failed with: %v", err)
+					glog.Errorf("UpdateNodeStatuses failed with: %v. Report volume %q as attached to node %q back",
```

Comment:
"UpdateNodeStatuses failed with: %v. Report volume %q as attached to node %q back" -> "UpdateNodeStatuses failed while attempting to report volume %q as attached to node %q with: %v "